sycl/test-e2e/KernelCompiler/kernel_compiler_sycl_jit.cpp

```diff
@@ -0,0 +1,129 @@
+    return a + b + 6;
+  }
+)===";
+
+// TODO: remove SYCL_EXTERNAL once it is no longer needed.
+auto constexpr SYCLSource = R"===(
+#include <sycl/sycl.hpp>
+#include "intermediate/AddEm.h"
+#include "intermediate/PlusEm.h"
+
+// use extern "C" to avoid name mangling
+extern "C" SYCL_EXTERNAL SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::ext::oneapi::experimental::nd_range_kernel<1>))
+void ff_cp(int *ptr) {
+
+  // intentionally using deprecated routine, as opposed to this_work_item::get_nd_item<1>()
+  sycl::nd_item<1> Item = sycl::ext::oneapi::experimental::this_nd_item<1>();
+
+  sycl::id<1> GId = Item.get_global_id();
+  ptr[GId.get(0)] = AddEm(GId.get(0), 37);
+}
+)===";
+
+void test_1(sycl::queue &Queue, sycl::kernel &Kernel, int seed) {
+  constexpr int Range = 10;
+  int *usmPtr = sycl::malloc_shared<int>(Range, Queue);
+  int start = 3;
+
+  sycl::nd_range<1> R1{{Range}, {1}};
+
+  bool Passa = true;
+
+  memset(usmPtr, 0, Range * sizeof(int));
+  Queue.submit([&](sycl::handler &Handler) {
+    Handler.set_arg(0, usmPtr);
+    Handler.parallel_for(R1, Kernel);
+  });
+  Queue.wait();
+
+  for (int i = 0; i < Range; i++) {
+    std::cout << usmPtr[i] << "=" << (i + seed) << " ";
+    assert(usmPtr[i] == i + seed);
+  }
+  std::cout << std::endl;
+
+  sycl::free(usmPtr, Queue);
+}
+
+void test_build_and_run() {
+  namespace syclex = sycl::ext::oneapi::experimental;
+  using source_kb = sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source>;
+  using exe_kb = sycl::kernel_bundle<sycl::bundle_state::executable>;
+
+  sycl::queue q;
+  sycl::context ctx = q.get_context();
+
+  bool ok =
+      q.get_device().ext_oneapi_can_compile(syclex::source_language::sycl_jit);
+  if (!ok) {
+    std::cout << "Apparently this device does not support `sycl_jit` source "
+                 "kernel bundle extension: "
+              << q.get_device().get_info<sycl::info::device::name>()
+              << std::endl;
+    return;
+  }
+
+  // Create from source.
+  syclex::include_files incFiles{"intermediate/AddEm.h", AddEmH};
+  incFiles.add("intermediate/PlusEm.h", PlusEmH);
+  source_kb kbSrc = syclex::create_kernel_bundle_from_source(
+      ctx, syclex::source_language::sycl_jit, SYCLSource,
+      syclex::properties{incFiles});
+
+  // Double check kernel_bundle.get_source() / get_backend().
+  sycl::context ctxRes = kbSrc.get_context();
+  assert(ctxRes == ctx);
+  sycl::backend beRes = kbSrc.get_backend();
+  assert(beRes == ctx.get_backend());
+
+  // Compilation of empty prop list, no devices.
+  exe_kb kbExe1 = syclex::build(kbSrc);
+
+  // // Compilation with props and devices
+  std::string log;
+  std::vector<std::string> flags{"-g", "-fno-fast-math"};
+  std::vector<sycl::device> devs = kbSrc.get_devices();
+  exe_kb kbExe2 = syclex::build(
+      kbSrc, devs, syclex::properties{syclex::build_options{flags}});
+
+  // extern "C" was used, so the name "ff_cp" is not mangled.
+  sycl::kernel k = kbExe2.ext_oneapi_get_kernel("ff_cp");
```

Comment:
My understanding was that we needed some different name for the kernel because of the way we outline kernel functions in the device compiler.  I thought the compiler generated a wrapper function around the user's kernel function which handles complex arguments that need to be decomposed.  For example, when an `accessor` is passed to a kernel, the device compiler decomposes this structure into its constituent member variables and passes each as a separate kernel argument.  The wrapper function reassembles the `accessor` from these separate kernel arguments and then calls the user's kernel function.

Since the wrapper function and the user's kernel function can't both have the same name, we added a prefix to the wrapper function.

How do you solve this problem if the kernel does not have some prefix?

Reply:
Are you referring to the `__sycl_kernel` prefix? That's currently special-handled transparently here: https://github.com/intel/llvm/blob/407124bdda90e230347fa2d8f14cc932585293b0/sycl/source/detail/kernel_bundle_impl.hpp#L508-L517

Reply:
Yes, that makes sense.  Thanks!